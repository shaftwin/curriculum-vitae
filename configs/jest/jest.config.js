const path = require('path');

module.exports = {
  rootDir: '../../',
  testEnvironment: 'jsdom',
  testMatch: ['**/*.test.js'],
  setupFilesAfterEnv: ['./node_modules/@testing-library/jest-dom'],
  moduleDirectories: [path.resolve(__dirname, '../../'), 'node_modules'],
  setupFiles: [
    './configs/jest/mock.js',
    './configs/cucumber/jest-cucumber-config.js',
  ],
  snapshotSerializers: ['@emotion/jest/serializer'],
};
