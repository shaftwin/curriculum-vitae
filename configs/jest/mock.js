jest.mock('react-i18next', () => ({
  withTranslation: () => (Component) => {
    // eslint-disable-next-line no-param-reassign
    Component.defaultProps = { ...Component.defaultProps, t: () => '' };
    return Component;
  },
  // this mock makes sure any components using the
  // translate hook can use it without a warning being shown
  useTranslation: () => {
    return {
      t: (str) => str,
      i18n: {
        changeLanguage: () => new Promise(() => {}),
      },
    };
  },
}));
