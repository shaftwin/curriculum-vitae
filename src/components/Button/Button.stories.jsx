/* eslint-disable react/jsx-props-no-spreading */
import React from 'react';
// eslint-disable-next-line import/no-extraneous-dependencies
import { actions } from '@storybook/addon-actions';
import Filter from '../../../assets/svg/filter';
import Button from './index';

/* More on default export:
https://storybook.js.org/docs/react/writing-stories/introduction#default-export
 */
export default {
  title: 'Components/Button',
  component: Button,
};

/* More on component templates:
https://storybook.js.org/docs/react/writing-stories/introduction#using-args
*/

const Template = (args) => <Button onclick={actions('onClick')} {...args} />;

// More on args: https://storybook.js.org/docs/react/writing-stories/args

export const FillNormal = Template.bind({});
FillNormal.args = {
  label: 'Button',
};

export const FillWithIcon = Template.bind({});
FillWithIcon.args = {
  icon: <Filter />,
  label: 'Button',
};

export const StrokeNormal = Template.bind({});
StrokeNormal.args = {
  label: 'Button',
  buttonType: 'stroke',
};

export const StrokeWithIcon = Template.bind({});
StrokeWithIcon.args = {
  icon: <Filter />,
  label: 'Button',
  buttonType: 'stroke',
};

export const SmallWithIcon = Template.bind({});
SmallWithIcon.args = {
  icon: <Filter />,
  label: 'Button',
  size: 'small',
};
