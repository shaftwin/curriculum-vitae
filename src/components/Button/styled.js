import styled from '@emotion/styled';

export const StyledButton = styled.button`
  display: flex;
  flex-direction: ${({ iconPosition }) =>
    iconPosition === 'left' ? 'row' : 'row-reverse'};
  align-items: center;
  justify-content: space-around;
  min-width: ${({ icon }) => (icon ? '7.5rem' : '5.75rem')};
  height: 2rem;
  border-radius: 0.438rem;
  cursor: pointer;
  font-weight: ${({ theme }) => theme.global.fontWeight.bold};
  font-size: ${({ theme, size }) => theme.button.fontSize[size]};
  background-color: ${({ theme, templateColor, buttonType }) =>
    theme.button.background[templateColor].default[buttonType]};
  border: ${({ theme, templateColor, buttonType }) =>
    buttonType === 'stroke'
      ? `solid 2px ${theme.button.borderColor[templateColor].default}`
      : 0};
  color: ${({ theme, templateColor }) =>
    theme.button.textColor[templateColor].default};

  &:hover {
    background-color: ${({ theme, templateColor, buttonType }) =>
      theme.button.background[templateColor].hover[buttonType]};
    border: ${({ theme, templateColor, buttonType }) =>
      buttonType === 'stroke'
        ? `solid 2px ${theme.button.borderColor[templateColor].hover}`
        : 0};
    color: ${({ theme, templateColor }) =>
      theme.button.textColor[templateColor].hover};
  }

  &:disabled {
    cursor: default;
    background-color: ${({ theme, templateColor, buttonType }) =>
      theme.button.background[templateColor].disabled[buttonType]};
    border: ${({ theme, templateColor, buttonType }) =>
      buttonType === 'stroke'
        ? `solid 2px ${theme.button.borderColor[templateColor].disabled}`
        : 0};
    color: ${({ theme, templateColor }) =>
      theme.button.textColor[templateColor].disabled};
  }
`;

export const IconContainer = styled.div`
  margin-top: 4px;
`;
