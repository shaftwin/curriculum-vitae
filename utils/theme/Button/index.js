import { global } from '../global';

export const button = {
  background: {
    primary: {
      default: {
        fill: global.colors.green,
        stroke: global.colors.white,
        text: global.colors.white,
      },
      hover: {
        fill: global.colors.hover.green,
        stroke: global.colors.white,
        text: global.colors.white,
      },
      disabled: {
        fill: global.colors.grey,
        stroke: global.colors.white,
        text: global.colors.white,
      },
    },
    secondary: {
      default: {
        fill: global.colors.lightGreen,
        stroke: global.colors.white,
        text: global.colors.white,
      },
      hover: {
        fill: global.colors.hover.lightGreen,
        stroke: global.colors.white,
        text: global.colors.white,
      },
      disabled: {
        fill: global.colors.grey,
        stroke: global.colors.white,
        text: global.colors.white,
      },
    },
    tertiary: {
      default: {
        fill: global.colors.darkBlue,
        stroke: global.colors.white,
      },
      hover: {
        fill: global.colors.hover.green,
        stroke: global.colors.white,
      },
      disabled: {
        fill: global.colors.grey,
        stroke: global.colors.white,
      },
    },
    white: {
      default: {
        fill: global.colors.white,
        stroke: global.colors.green,
        text: global.colors.green,
      },
      hover: {
        fill: global.colors.lightGreen,
        stroke: global.colors.green,
        text: global.colors.green,
      },
    },
  },
  textColor: {
    primary: {
      default: global.colors.darkBlue,
      hover: global.colors.darkBlue,
      disabled: global.colors.disabledGrey,
    },
    secondary: {
      default: global.colors.darkBlue,
      hover: global.colors.darkBlue,
      disabled: global.colors.disabledGrey,
    },
    tertiary: {
      default: global.colors.white,
      hover: global.colors.white,
      disabled: global.colors.disabledGrey,
    },
    white: {
      default: global.colors.darkBlue,
      hover: global.colors.hover.darkBlue,
      disabled: global.colors.brownGrey,
    },
  },
  // We use the border only when the type is stroke
  borderColor: {
    primary: {
      default: global.colors.green,
      hover: global.colors.hover.green,
      disabled: global.colors.grey,
    },
    secondary: {
      default: global.colors.lightGreen,
      hover: global.colors.hover.lightGreen,
      disabled: global.colors.grey,
    },
    tertiary: {
      default: global.colors.darkBlue,
      hover: global.colors.hover.green,
      disabled: global.colors.grey,
    },
    white: {
      default: global.colors.white,
      hover: global.colors.lightGreen,
      disabled: global.colors.grey,
    },
  },
  fontSize: {
    extraSmall: global.fontSize[10],
    small: global.fontSize[12],
    medium: global.fontSize[13],
  },
};
