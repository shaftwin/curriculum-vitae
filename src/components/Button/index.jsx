import React from 'react';
import PropTypes from 'prop-types';
import { StyledButton, IconContainer } from './styled';

const Button = ({
  label,
  size,
  onClick,
  disabled,
  icon,
  iconPosition,
  className,
  templateColor,
  buttonType,
  ...props
}) => {
  return (
    <StyledButton
      className={className}
      disabled={disabled}
      size={size}
      onClick={onClick}
      icon={icon}
      iconPosition={iconPosition}
      templateColor={templateColor}
      buttonType={buttonType}
      // eslint-disable-next-line react/jsx-props-no-spreading
      {...props}
    >
      {!!icon && (
        <IconContainer data-testid="iconContainer">{icon}</IconContainer>
      )}
      {label}
    </StyledButton>
  );
};

Button.propTypes = {
  /**
   * Button label
   */
  label: PropTypes.string.isRequired,
  /**
   * Optional button size
   */
  size: PropTypes.oneOf(['extraSmall', 'small', 'medium']),
  /**
   * Optional click handler
   */
  onClick: PropTypes.func,
  /**
   * Optional disabled
   */
  disabled: PropTypes.bool,
  /**
   * Optional icon
   */
  icon: PropTypes.element,
  /**
   * Optional iconPosition
   */
  iconPosition: PropTypes.oneOf(['left', 'right']),
  /**
   * Optional custom style
   */
  className: PropTypes.string,
  /**
   * Optional template color
   */
  templateColor: PropTypes.oneOf(['primary', 'secondary', 'tertiary', 'white']),
  /**
   * Optional buttonType
   */
  buttonType: PropTypes.oneOf(['fill', 'stroke', 'text']),
};

Button.defaultProps = {
  onClick: () => null,
  disabled: false,
  icon: null,
  iconPosition: 'left',
  size: 'medium',
  templateColor: 'primary',
  buttonType: 'fill',
};

export default Button;
