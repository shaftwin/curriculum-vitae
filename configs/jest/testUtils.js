// import { MockedProvider } from '@apollo/client/testing';
import { ThemeProvider } from '@emotion/react';
import { render } from '@testing-library/react';
import React from 'react';
import theme from '../../utils/theme';

const AllTheProviders = ({ children }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

const customRender = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
