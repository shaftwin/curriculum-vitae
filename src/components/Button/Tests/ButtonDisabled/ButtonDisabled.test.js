import * as React from 'react';
import { fireEvent, screen } from '@testing-library/react';

import { defineFeature, loadFeature } from 'jest-cucumber';
import { render } from '../../../../../configs/jest/testUtils';
import Button from '../..';

const feature = loadFeature(
  'src/components/Button/Tests/ButtonDisabled/ButtonDisabled.feature',
);

defineFeature(feature, (test) => {
  const onClick = jest.fn();

  test('Button clicked should not fire callBack method', ({
    given,
    when,
    then,
  }) => {
    let container;
    let button;
    given('Button is disabled', () => {
      container = render(
        <Button label="testLabel" disabled={true} onClick={onClick} />,
      ).container;
      expect(container).toMatchSnapshot();
      button = screen.getByText('testLabel');
    });

    when('Button is clicked', () => {
      fireEvent.click(button);
    });

    then('Button callback should not be called', () => {
      expect(onClick).toHaveBeenCalledTimes(0);
    });
  });
});
