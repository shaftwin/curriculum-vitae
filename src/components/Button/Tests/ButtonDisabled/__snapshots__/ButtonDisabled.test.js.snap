// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Disabled Button clicked should not fire callBack method 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: space-around;
  -ms-flex-pack: space-around;
  -webkit-justify-content: space-around;
  justify-content: space-around;
  min-width: 5.75rem;
  height: 2rem;
  border-radius: 0.438rem;
  cursor: pointer;
  font-weight: 700;
  font-size: 0.813rem;
  background-color: #00DBC7;
  border: 0;
  color: #120D2E;
}

.emotion-0:hover {
  background-color: #00C6B3;
  border: 0;
  color: #120D2E;
}

.emotion-0:disabled {
  cursor: default;
  background-color: #EBEEF0;
  border: 0;
  color: #888596;
}

<div>
  <button
    class="emotion-0"
    disabled=""
  >
    testLabel
  </button>
</div>
`;
